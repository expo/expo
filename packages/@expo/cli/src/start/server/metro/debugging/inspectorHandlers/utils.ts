import type WS from 'ws';

import { DebuggerResponse, DeviceResponse } from './types';

/**
 * Helper function to respond to a message from the debugger or device.
 * The return value is used to stop the message propagation, "canceling" further handling.
 *
 * @example ```
 *  return respond<DeviceResponse<CDP>>(socket, { id: message.id, result: {} });
 * ```
 */
export function respond<T = DeviceResponse | DebuggerResponse>(socket: WS, message: T) {
  socket.send(JSON.stringify(message));
  return true;
}

/** Known compatible debuggers that require specific workarounds */
export type DebuggerType = 'chrome' | 'vscode' | 'unknown';

// Patterns to test against user agents
const CHROME_USER_AGENT = /chrome/i;
const VSCODE_USER_AGENT = /vscode/i;

/**
 * Determine the debugger type based on the known user agent.
 */
export function getDebuggerType(userAgent?: string | null): DebuggerType {
  if (userAgent && CHROME_USER_AGENT.test(userAgent)) return 'chrome';
  if (userAgent && VSCODE_USER_AGENT.test(userAgent)) return 'vscode';
  return 'unknown';
}
