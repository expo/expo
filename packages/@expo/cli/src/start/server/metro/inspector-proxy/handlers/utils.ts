import type WS from 'ws';

import { DebuggerResponse, DeviceResponse } from './types';

/**
 * Helper function to respond to a message from the debugger or device.
 * The return value is used to stop the message propagation, "canceling" further handling.
 *
 * @example ```
 *  return respond<DeviceResponse<CDP>>(socket, { id: message.id, result: {} });
 * ```
 */
export function respond<T = DeviceResponse | DebuggerResponse>(socket: WS, message: T) {
  socket.send(JSON.stringify(message));
  return true;
}
