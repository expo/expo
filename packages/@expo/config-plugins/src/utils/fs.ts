import fs from 'fs';
import path from 'path';

/** A basic function that copies a single file to another file location. */
export async function copyFilePathToPathAsync(src: string, dest: string): Promise<void> {
  const srcFile = await fs.promises.readFile(src);
  await fs.promises.mkdir(path.dirname(dest), { recursive: true });
  await fs.promises.writeFile(dest, srcFile);
}

/** Remove a single file (not directory). Returns `true` if a file was actually deleted. */
export function removeFile(filePath: string): boolean {
  try {
    fs.unlinkSync(filePath);
    return true;
  } catch (error: any) {
    // Skip if the remove did nothing.
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}
