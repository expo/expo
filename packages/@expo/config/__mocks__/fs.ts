module.exports = require('memfs').fs;

// NOTE(cedric): workaround to also mock `node:fs`
jest.mock('node:fs', () => require('fs'));
