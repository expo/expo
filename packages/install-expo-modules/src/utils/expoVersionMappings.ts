import resolveFrom from 'resolve-from';
import semver from 'semver';

export interface VersionInfo {
  expoSdkVersion: string;
  iosDeploymentTarget: string;
  reactNativeVersionRange: string;
  androidAgpVersion?: string;
  supportCliIntegration?: boolean;
}

export const ExpoVersionMappings: VersionInfo[] = [
  // Please keep sdk versions in sorted order (latest sdk first)
  {
    expoSdkVersion: '50.0.0',
    iosDeploymentTarget: '13.4',
    reactNativeVersionRange: '>= 0.73.0',
    supportCliIntegration: true,
  },
  {
    expoSdkVersion: '49.0.0',
    iosDeploymentTarget: '13.0',
    reactNativeVersionRange: '>= 0.72.0',
    supportCliIntegration: true,
  },
  {
    expoSdkVersion: '48.0.0',
    iosDeploymentTarget: '13.0',
    reactNativeVersionRange: '>= 0.71.0',
    androidAgpVersion: '7.4.1',
  },
  {
    expoSdkVersion: '47.0.0',
    iosDeploymentTarget: '13.0',
    reactNativeVersionRange: '>= 0.70.0',
  },
  {
    expoSdkVersion: '46.0.0',
    iosDeploymentTarget: '12.4',
    reactNativeVersionRange: '>= 0.69.0',
  },
  {
    expoSdkVersion: '45.0.0',
    iosDeploymentTarget: '12.0',
    reactNativeVersionRange: '>= 0.65.0',
  },
  {
    expoSdkVersion: '44.0.0',
    iosDeploymentTarget: '12.0',
    reactNativeVersionRange: '< 0.68.0',
  },
  {
    expoSdkVersion: '43.0.0',
    iosDeploymentTarget: '12.0',
    reactNativeVersionRange: '< 0.68.0',
  },
];

export function getDefaultSdkVersion(projectRoot: string): VersionInfo {
  const reactNativePackageJsonPath = resolveFrom.silent(projectRoot, 'react-native/package.json');
  if (!reactNativePackageJsonPath) {
    throw new Error(`Unable to find react-native package - projectRoot[${projectRoot}]`);
  }
  const reactNativeVersion = require(reactNativePackageJsonPath).version;
  const versionInfo = ExpoVersionMappings.find((info) =>
    semver.satisfies(reactNativeVersion, info.reactNativeVersionRange)
  );
  if (!versionInfo) {
    throw new Error(
      `Unable to find compatible expo sdk version - reactNativeVersion[${reactNativeVersion}]`
    );
  }
  return versionInfo;
}

export function getLatestSdkVersion(): VersionInfo {
  return ExpoVersionMappings[0];
}

export function getVersionInfo(sdkVersion: string): VersionInfo | null {
  return ExpoVersionMappings.find((info) => info.expoSdkVersion === sdkVersion) ?? null;
}
