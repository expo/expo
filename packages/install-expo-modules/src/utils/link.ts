import chalk from 'chalk';
import terminalLink from 'terminal-link';

/**
 * Prints a link for given URL, using text if provided, otherwise text is just the URL.
 * Format links as dim (unless disabled) and with an underline.
 *
 * @example https://expo.dev
 */
export function link(
  url: string,
  { text = url, dim = true }: { text?: string; dim?: boolean } = {}
): string {
  let output: string;
  // Links can be disabled via env variables https://github.com/jamestalmage/supports-hyperlinks/blob/master/index.js
  if (terminalLink.isSupported) {
    output = terminalLink(text, url);
  } else {
    output = `${text === url ? '' : text + ': '}${chalk.underline(url)}`;
  }
  return dim ? chalk.dim(output) : output;
}

/**
 * Provide a consistent "Learn more" link experience.
 * Format links as dim (unless disabled) with an underline.
 *
 * @example [Learn more](https://expo.dev)
 * @example Learn more: https://expo.dev
 */
export function learnMore(
  url: string,
  {
    learnMoreMessage: maybeLearnMoreMessage,
    dim = true,
  }: { learnMoreMessage?: string; dim?: boolean } = {}
): string {
  return link(url, { text: maybeLearnMoreMessage ?? 'Learn more', dim });
}
