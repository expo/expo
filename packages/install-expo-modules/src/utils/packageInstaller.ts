import * as PackageManager from '@expo/package-manager';
import path from 'path';
import semver from 'semver';

/**
 * Install a node package non-interactively
 *
 * @param projectRoot target project root folder
 * @param pkg package name
 */
function installPackageNonInteractiveAsync(projectRoot: string, pkg: string) {
  const manager = PackageManager.createForProject(projectRoot);

  switch (manager.name) {
    case 'yarn':
      return manager.addAsync([pkg, '--non-interactive']);

    case 'bun':
    case 'npm':
    case 'pnpm':
      return manager.addAsync([pkg]);
  }
}

/**
 * Install the `expo` package
 *
 * @param projectRoot target project root folder
 * @param sdkVersion expo sdk version
 */
export async function installExpoPackageAsync(projectRoot: string, sdkVersion: string) {
  try {
    // First try to install from released versions, e.g. `expo@^45.0.0`
    await installPackageNonInteractiveAsync(projectRoot, `expo@^${sdkVersion}`);
  } catch {
    // Fallback to install from prerelease versions,
    // e.g. `expo@>=45.0.0-0 <46.0.0`, this will cover prerelease version for beta testing.
    await installPackageNonInteractiveAsync(
      projectRoot,
      `expo@>=${sdkVersion}-0 <${semver.inc(sdkVersion, 'major')}`
    );
  }
}

/**
 * Running `pod install` for the target project
 *
 * @param projectRoot target project root folder
 */
export async function installPodsAsync(projectRoot: string) {
  const podPackageManager = new PackageManager.CocoaPodsPackageManager({
    cwd: path.join(projectRoot, 'ios'),
  });
  await podPackageManager.installAsync();
}
