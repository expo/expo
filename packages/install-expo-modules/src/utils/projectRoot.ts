import assert from 'assert';
import findUp from 'find-up';
import path from 'path';

function findUpPackageJson(root: string): string {
  const packageJson = findUp.sync('package.json', { cwd: root });
  assert(packageJson, `No package.json found for module "${root}"`);
  return packageJson;
}

export function normalizeProjectRoot(projectRoot?: string): string {
  return path.dirname(findUpPackageJson(projectRoot ?? process.cwd()));
}
