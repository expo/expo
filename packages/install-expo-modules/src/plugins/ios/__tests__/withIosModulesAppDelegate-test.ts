import fs from 'fs';
import path from 'path';

import { getLatestSdkVersion } from '../../../utils/expoVersionMappings';
import {
  updateModulesAppDelegateObjcHeader,
  updateModulesAppDelegateObjcImpl,
  updateModulesAppDelegateSwift,
} from '../withIosModulesAppDelegate';

const fixturesPath = path.resolve(__dirname, 'fixtures');

describe(updateModulesAppDelegateObjcHeader, () => {
  it('should migrate from react-native@>=0.71.0 AppDelegate header', async () => {
    const [rawContents, expectContents] = await Promise.all([
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-rn071.h'), 'utf8'),
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-rn071-updated.h'), 'utf8'),
    ]);

    const sdkVersion = getLatestSdkVersion().expoSdkVersion;
    const contents = updateModulesAppDelegateObjcHeader(rawContents, sdkVersion);
    expect(updateModulesAppDelegateObjcHeader(contents, sdkVersion)).toEqual(expectContents);
    // Try it twice...
    const nextContents = updateModulesAppDelegateObjcHeader(contents, sdkVersion);
    expect(updateModulesAppDelegateObjcHeader(nextContents, sdkVersion)).toEqual(expectContents);
  });

  it('should migrate from classic RN AppDelegate header', async () => {
    const [rawContents, expectContents] = await Promise.all([
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate.h'), 'utf8'),
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-updated.h'), 'utf8'),
    ]);

    const sdkVersion = getLatestSdkVersion().expoSdkVersion;
    const contents = updateModulesAppDelegateObjcHeader(rawContents, sdkVersion);
    expect(updateModulesAppDelegateObjcHeader(contents, sdkVersion)).toEqual(expectContents);
    // Try it twice...
    const nextContents = updateModulesAppDelegateObjcHeader(contents, sdkVersion);
    expect(updateModulesAppDelegateObjcHeader(nextContents, sdkVersion)).toEqual(expectContents);
  });
});

describe(updateModulesAppDelegateObjcImpl, () => {
  it('should migrate from classic react-native@>=0.68.0 AppDelegate.mm implementation', async () => {
    const [rawContents, expectContents] = await Promise.all([
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-rn068.mm'), 'utf8'),
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-rn068-updated.mm'), 'utf8'),
    ]);

    const sdkVersion = getLatestSdkVersion().expoSdkVersion;
    const contents = updateModulesAppDelegateObjcImpl(rawContents, sdkVersion);
    expect(contents).toEqual(expectContents);
    // Try it twice...
    const nextContents = updateModulesAppDelegateObjcImpl(contents, sdkVersion);
    expect(nextContents).toEqual(expectContents);
  });

  it('should migrate from classic react-native@<0.68.0 AppDelegate.m implementation', async () => {
    const [rawContents, expectContents] = await Promise.all([
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate.m'), 'utf8'),
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-updated.m'), 'utf8'),
    ]);

    const sdkVersion = getLatestSdkVersion().expoSdkVersion;
    const contents = updateModulesAppDelegateObjcImpl(rawContents, sdkVersion);
    expect(contents).toEqual(expectContents);
    // Try it twice...
    const nextContents = updateModulesAppDelegateObjcImpl(contents, sdkVersion);
    expect(nextContents).toEqual(expectContents);
  });
});

describe(updateModulesAppDelegateSwift, () => {
  it('should migrate from basic RN AppDelegate.swift', async () => {
    const [rawContents, expectContents] = await Promise.all([
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate.swift'), 'utf8'),
      fs.promises.readFile(path.join(fixturesPath, 'AppDelegate-updated.swift'), 'utf8'),
    ]);

    const sdkVersion = getLatestSdkVersion().expoSdkVersion;
    expect(updateModulesAppDelegateSwift(rawContents, sdkVersion)).toEqual(expectContents);
  });
});

describe('withIosModulesAppDelegate sdkVersion snapshots', () => {
  const [objcHeaderFixture, objcImplFixture, swiftFixture] = [
    fs.readFileSync(path.join(fixturesPath, 'AppDelegate.h'), 'utf8'),
    fs.readFileSync(path.join(fixturesPath, 'AppDelegate.m'), 'utf8'),
    fs.readFileSync(path.join(fixturesPath, 'AppDelegate.swift'), 'utf8'),
  ];

  ['43.0.0', '44.0.0'].forEach((sdkVersion) => {
    it(`sdkVersion ${sdkVersion}`, () => {
      expect(updateModulesAppDelegateObjcHeader(objcHeaderFixture, sdkVersion)).toMatchSnapshot();
      expect(updateModulesAppDelegateObjcImpl(objcImplFixture, sdkVersion)).toMatchSnapshot();
      expect(updateModulesAppDelegateSwift(swiftFixture, sdkVersion)).toMatchSnapshot();
    });
  });
});
