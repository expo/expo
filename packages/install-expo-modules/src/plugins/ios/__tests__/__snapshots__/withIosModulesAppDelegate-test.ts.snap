// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withIosModulesAppDelegate sdkVersion snapshots sdkVersion 43.0.0 1`] = `
"#import <React/RCTBridgeDelegate.h>
#import <Expo/Expo.h>
#import <UIKit/UIKit.h>

@interface AppDelegate : EXAppDelegateWrapper <UIApplicationDelegate, RCTBridgeDelegate>

@property (nonatomic, strong) UIWindow *window;

@end
"
`;

exports[`withIosModulesAppDelegate sdkVersion snapshots sdkVersion 43.0.0 2`] = `
"#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
#ifdef FB_SONARKIT_ENABLED
  InitializeFlipper(application);
#endif

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                               moduleName:@"RN0633"
                                               initialProperties:nil];

  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [super application:application didFinishLaunchingWithOptions:launchOptions];
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

@end
"
`;

exports[`withIosModulesAppDelegate sdkVersion snapshots sdkVersion 43.0.0 3`] = `
"import UIKit

@UIApplicationMain
class AppDelegate: AppDelegateWrapper, RCTBridgeDelegate {
  var window: UIWindow?

  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    let bridge = RCTBridge(delegate: self, launchOptions: launchOptions)!
    let rootView = RCTRootView(bridge: bridge, moduleName: "HelloWorld", initialProperties: nil)

    if #available(iOS 13.0, *) {
      rootView.backgroundColor = UIColor.systemBackground
    } else {
      rootView.backgroundColor = UIColor.white
    }

    self.window = UIWindow(frame: UIScreen.main.bounds);
    let rootViewController = UIViewController()
    rootViewController.view = rootView
    self.window?.rootViewController = rootViewController
    self.window?.makeKeyAndVisible()
    super.application(application, didFinishLaunchingWithOptions: launchOptions)
    return true
  }

  func sourceURL(for bridge: RCTBridge!) -> URL! {
    #if DEBUG
    return RCTBundleURLProvider.sharedSettings()?.jsBundleURL(forBundleRoot: "index", fallbackResource: nil)
    #else
    return Bundle.main.url(forResource: "main", withExtension: "jsBundle")
    #endif
  }
}
"
`;

exports[`withIosModulesAppDelegate sdkVersion snapshots sdkVersion 44.0.0 1`] = `
"#import <React/RCTBridgeDelegate.h>
#import <Expo/Expo.h>
#import <UIKit/UIKit.h>

@interface AppDelegate : EXAppDelegateWrapper <UIApplicationDelegate, RCTBridgeDelegate>

@property (nonatomic, strong) UIWindow *window;

@end
"
`;

exports[`withIosModulesAppDelegate sdkVersion snapshots sdkVersion 44.0.0 2`] = `
"#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
#ifdef FB_SONARKIT_ENABLED
  InitializeFlipper(application);
#endif

  RCTBridge *bridge = [self.reactDelegate createBridgeWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [self.reactDelegate createRootViewWithBridge:bridge
                                               moduleName:@"RN0633"
                                               initialProperties:nil];

  rootView.backgroundColor = [[UIColor alloc] initWithRed:1.0f green:1.0f blue:1.0f alpha:1];

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [self.reactDelegate createRootViewController];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [super application:application didFinishLaunchingWithOptions:launchOptions];
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}

@end
"
`;

exports[`withIosModulesAppDelegate sdkVersion snapshots sdkVersion 44.0.0 3`] = `
"import UIKit

@UIApplicationMain
class AppDelegate: AppDelegateWrapper, RCTBridgeDelegate {
  var window: UIWindow?

  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    let bridge = reactDelegate.createBridge(delegate: self, launchOptions: launchOptions)!
    let rootView = reactDelegate.createRootView(bridge: bridge, moduleName: "HelloWorld", initialProperties: nil)

    if #available(iOS 13.0, *) {
      rootView.backgroundColor = UIColor.systemBackground
    } else {
      rootView.backgroundColor = UIColor.white
    }

    self.window = UIWindow(frame: UIScreen.main.bounds);
    let rootViewController = reactDelegate.createRootViewController()
    rootViewController.view = rootView
    self.window?.rootViewController = rootViewController
    self.window?.makeKeyAndVisible()
    super.application(application, didFinishLaunchingWithOptions: launchOptions)
    return true
  }

  func sourceURL(for bridge: RCTBridge!) -> URL! {
    #if DEBUG
    return RCTBundleURLProvider.sharedSettings()?.jsBundleURL(forBundleRoot: "index", fallbackResource: nil)
    #else
    return Bundle.main.url(forResource: "main", withExtension: "jsBundle")
    #endif
  }
}
"
`;
